lib/workers.js

```diff
@@ -0,0 +1,222 @@
+  test = Object.assign(new Test(test.title || '', () => { }), test);
+  test.parent = Object.assign(new Suite(test.parent.title), test.parent);
+  return test;
+};
+
+const createWorkerObjects = (testGroups, config) => {
+  return testGroups.map((tests, index) => new WorkerObject({ config: JSON.stringify(config) }, tests, getTestRoot(), index));
+};
+
+const indexOfSmallestElement = (groups) => {
+  let i = 0;
+  for (let j = 1; j < groups.length; j++) {
+    if (groups[j - 1].length > groups[j].length) {
+      i = j;
+    }
+  }
+  return i;
+};
+
+class WorkerObject {
+  constructor(options, tests, testRoot, workerIndex) {
+    this.options = options || {};
+    this.tests = tests;
+    this.testRoot = testRoot || getTestRoot();
+    this.workerIndex = workerIndex;
+  }
+
+  addConfig(config) {
+    this.options.config = JSON.stringify(config);
+  }
+
+  addTests(tests) {
+    this.tests = tests;
+  }
+}
+
+class Workers extends EventEmitter {
+  /**
+   * @param {Number} numberOfWorkers
+   * @param {Boolean} runAsSuite
+   */
+  constructor(numberOfWorkers, runAsSuite, configPath) {
+    super();
+    this.codecept = initializeCodecept(configPath);
+    this.finishedTests = [];
+    this.errors = [];
+    this.numberOfWorkers = 0;
+    this.closedWorkers = 0;
+    this.workers = [];
+    this.stats = {
+      passes: 0,
+      failures: 0,
+      tests: 0,
+    };
+
+    if (typeof numberOfWorkers === 'number' && numberOfWorkers > 0) {
+      this.numberOfWorkers = numberOfWorkers;
+      this.testGroups = runAsSuite ? this.createGroupsOfSuites(numberOfWorkers) : this.createGroupsOfTests(numberOfWorkers);
+      this.workers = createWorkerObjects(this.testGroups, this.codecept.config);
+    }
+  }
+
+  spawn() {
+    const worker = new WorkerObject({}, [], getTestRoot(), this.numberOfWorkers);
+    this.workers.push(worker);
+    this.numberOfWorkers += 1;
+    return worker;
+  }
+
+  createGroupsOfTests(numberOfWorkers) {
+    const files = this.codecept.testFiles;
+    const mocha = Container.mocha();
+    mocha.files = files;
+    mocha.loadFiles();
+    const groups = populateGroups(numberOfWorkers);
+    let groupCounter = 0;
+
+    mocha.suite.eachTest((test) => {
+      const i = groupCounter % groups.length;
+      if (test) {
+        const { id } = test;
+        groups[i].push(id);
+        groupCounter++;
+      }
+    });
+    return groups;
+  }
+
+  createGroupsOfSuites(numberOfWorkers) {
+    const files = this.codecept.testFiles;
+    const mocha = Container.mocha();
+    mocha.files = files;
+    mocha.loadFiles();
+    const groups = populateGroups(numberOfWorkers);
+
+    mocha.suite.suites.forEach((suite) => {
+      const i = indexOfSmallestElement(groups);
+      suite.tests.forEach((test) => {
+        if (test) {
+          const { id } = test;
+          groups[i].push(id);
+        }
+      });
+    });
+    return groups;
+  }
+
+  run() {
+    this.codecept.runBootstrap((err) => {
```

Comment:
worker should execute `bootstrapAll` instead of bootstrap.

See 

```
  runHook(config.bootstrapAll, () => {
    const workerList = createWorkers(groups, options, testRoot);
    workerList.forEach(worker => assignWorkerMethods(worker, groups.length));
  }, 'bootstrapAll');
```
runworkers.js:63